### ISSUE TITLE
Type safety feature requests collection

### ISSUE THREAD

**Issue by Perryvw** [tangzx/IntelliJ-EmmyLua#60]
Version 0.1 of the type safety setting was completed and can be found in pull request #59. 

However, the feature is not done yet, some things we would still like to add:

- Return type checking
- Generic functions e.g. fun<T>(in: T[]): T

Additionally, there are currently some workarounds in the code to negate some missing features of the original plugin functionality.

This is a list of features we would like to see to continue developing the type safety sytem:

- [ ] Parse 'any' in docs as TyUnknown (issue #58)
- [ ] Parse 'nil' in code as TyNil type instead of TyUnknown
- [ ] Add `--- @override` annotation that copies the doc of the function from the closest superclass.

Additionally there are some more complex features we would like to see eventually, but are not really a priority right now:

- [ ] Allow for generic function specification in docs, something like `--- @generic T`, `--- @param x T`, `---
 @returns T[]`
- [ ] Some method to append to the docs of already specified functions (for example in the standard library). I am not sure how this would look exactly.

More issues might arise later, but for now having the top list would help us out greatly.

**Comment by Perryvw** [comment]
Figured out how to do parsing of nil and primitives into their proper class types, updated top comment.

**Comment by adriweb** [comment]
I haven't looked in detail yet, but are userdata and thread (coroutines) types handled by that?
Anyway, good job, it looks nice :)

**Comment by Perryvw** [comment]
I did not take userdata and thread into account, I don't have any experience or testcases with those in lua. I looked them up in the lua docs and they are official types, so I think they should also be considered primitives, would you agree?

**Comment by adriweb** [comment]
Yep, if possible :)

**Comment by Perryvw** [comment]
Will add them later today!

**Comment by tangzx** [comment]
No, I think userdata is just a runtime type, it also can be a defined doc-type, not same as primitives

**Comment by adriweb** [comment]
I suppose that as long as it works with type annotations, it's fine for me.
Technically at "compile"-time (I mean, we know the type before execution, anyway), a userdata can be obtained with `newproxy()`. But indeed it's not something like `"foobar"` that's directly of type string.
Same goes for threads when creating coroutines.

**Comment by Perryvw** [comment]
According to https://www.lua.org/pil/2.html userdata and thread are proper lua types, therefore I would say implement them the same we I have implemented table and function primitives, that serve as a kind of 'any' type for tables and functions.

I.e. `fun(a: number): string` is a subtype of 'function' (TyPrimitive(PrimitiveKind.Function))

The system allows more specific types to be defined with these as superclass.

**Comment by Perryvw** [comment]
After thinking about this for a day I have to side with @tangzx on this, I will try to explain my reasoning:

While it is true that `thread` and `userdata` are considered as their own type in lua, there are (as far as I know) no special ways to declare them, i.e. `function() end` for type function, `{}` for type table. Every instance of these types will therefore originate from some form of API call. Therefore considering these two types as just regular classes will suffice in this case, assuming the methods spawning them are properly annotated (spoiler: the standard library is NOT properly annotated for this).

Therefore I think that for now `thread` and `userdata` will be considered just instances of the class type, unless a good reason is found not to do so.

**Comment by DoctorGester** [comment]
A good feature would be prioritizing variables of required type in autocomplete when typing an argument

**Comment by fisker583** [comment]
#
